File: code/matlab/apply_cloud_kernels_v2.m
% This script demonstrates how to apply the cloud radiative kernels to
% estimate the radiative impact of ISCCP simulator-defined cloud fraction anomalies 
% Dividing these radiative impacts by the global mean surface temperature
% anomaly yields an estimate of the cloud feedback

% Reference: Zelinka, Mark D., Stephen A. Klein, Dennis L. Hartmann, 2012: Computing and Partitioning Cloud Feedbacks Using 
% Cloud Property Histograms. Part I: Cloud Radiative Kernels. J. Climate, 25, 3715?3735. doi:10.1175/JCLI-D-11-00248.1.

% v2: This script is written to demonstrate how to compute the cloud feedback using for a 
% short (2-year) period of MPI-ESM-LR using the difference between amipFuture and amip runs.
% One should difference longer periods for more robust results -- these are just for demonstrative purposes

% Data that are used in this script:
% 1. model clisccp field
% 2. model rsuscs field
% 3. model rsdscs field
% 4. model tas field
% 5. cloud radiative kernels

% Other scripts that are called by this script:
% 1. get_netcdf_data.m
% 2. count_wt_mean.m

% Lines that include the comment "USER: MODIFY THIS LINE" are those in which 
% the user must update the path to the relevant netcdf files

% This script written by Mark Zelinka (zelinka1@llnl.gov) on 23 June 2014
% v2 modifications done on 13 July 2017

%% Load in the cloud radiative kernels
% Kernels are in units of W/m2/%
% Kernels are size (mo,tau,CTP,lat,albcs)

variable='/g/g19/zelinka1/CMIP3_kernels/cloud_kernels2.nc'; % USER: MODIFY THIS LINE 
LWkernel = get_netcdf_data(variable,'LWkernel'); 
SWkernel = get_netcdf_data(variable,'SWkernel'); 
albcs_midpt = get_netcdf_data(variable,'albcs'); % 0, 0.5, 1.0
kern_lat = get_netcdf_data(variable,'lat');  
kern_lon=1.25:2.5:360;
kern_coslat=cos(pi*kern_lat./180);

%% Load model clisccp, rsuscs, rsutcs, tas from amip control run 
variable='/p/lscratche/zelinka1/cmip5/clisccp/clisccp_cfMon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
ctl_clisccp = get_netcdf_data(variable,'clisccp'); % size (time,tau,CTP,lat,lon); units: percent
ctl_clisccp(ctl_clisccp>500)=NaN; 

variable='/p/lscratche/zelinka1/cmip5/tas/tas_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
ctl_tas = get_netcdf_data(variable,'tas'); % size (time,lat,lon)
ctl_tas(ctl_tas>500)=NaN; 

variable='/p/lscratche/zelinka1/cmip5/rsdscs/rsdscs_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
ctl_rsdscs = get_netcdf_data(variable,'rsdscs'); % size (time,lat,lon)
ctl_rsdscs(ctl_rsdscs>500)=NaN; 

variable='/p/lscratche/zelinka1/cmip5/rsuscs/rsuscs_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
ctl_rsuscs = get_netcdf_data(variable,'rsuscs'); % size (time,lat,lon)
ctl_rsuscs(ctl_rsuscs>500)=NaN; 

lon = get_netcdf_data(variable,'lon'); % size (time,lat,lon)
lat = get_netcdf_data(variable,'lat'); % size (time,lat,lon)
coslat=cos(pi*lat./180);

% Compute clear-sky surface albedo
ctl_albcs=ctl_rsuscs./ctl_rsdscs; 
clear ctl_rsuscs ctl_rsdscs

%% Load model clisccp and tas from amipFuture run 
variable='/p/lscratche/zelinka1/cmip5/clisccp/clisccp_cfMon_MPI-ESM-LR_amipFuture_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
fut_clisccp = get_netcdf_data(variable,'clisccp'); % size (time,tau,CTP,lat,lon); units: percent
fut_clisccp(fut_clisccp>500)=NaN; 

variable='/p/lscratche/zelinka1/cmip5/tas/tas_Amon_MPI-ESM-LR_amipFuture_r1i1p1_197901-198112.nc'; % USER: MODIFY THIS LINE 
fut_tas = get_netcdf_data(variable,'tas'); % size (time,lat,lon)
fut_tas(fut_tas>500)=NaN; 

%% Compute anual cycles
[a,b,c,d,e]=size(ctl_clisccp);
avgctl_albcs=nan*ones(12,d,e);
avgctl_tas=nan*ones(12,d,e);
avgfut_tas=nan*ones(12,d,e);
avgctl_clisccp=nan*ones(12,b,c,d,e);
avgfut_clisccp=nan*ones(12,b,c,d,e);
% before blindly doing this for every model, make sure your model data start in January
for M=1:12
    avgctl_clisccp(M,:,:,:,:)=squeeze(nanmean(ctl_clisccp(M:12:end,:,:,:,:),1));
    avgfut_clisccp(M,:,:,:,:)=squeeze(nanmean(fut_clisccp(M:12:end,:,:,:,:),1));
    avgctl_tas(M,:,:)=squeeze(nanmean(ctl_tas(M:12:end,:,:),1));
    avgfut_tas(M,:,:)=squeeze(nanmean(fut_tas(M:12:end,:,:),1));
    avgctl_albcs(M,:,:)=squeeze(nanmean(ctl_albcs(M:12:end,:,:),1));
end
clear ctl_clisccp fut_clisccp ctl_tas fut_tas ctl_albcs

%% Difference the anual cycles 
anom_clisccp=avgfut_clisccp-avgctl_clisccp;
anom_tas=avgfut_tas-avgctl_tas;
clear avgfut_clisccp avgctl_clisccp avgfut_tas avgctl_tas

%% Compute global mean surface air temperature anomaly
avgdtas=count_wt_mean(squeeze(nanmean(nanmean(anom_tas,1),3)),coslat',2);
clear anom_tas

%% Interpolate to a common grid (use the same grid as the kernels)
% first, a kluge to deal with weirdness of Matlab's interp function
x=[lon;lon(1)+360];
cat_albcs=cat(3,avgctl_albcs,avgctl_albcs(:,:,1)); 
cat_clisccp=cat(5,anom_clisccp,anom_clisccp(:,:,:,:,1)); 
clear anom_clisccp avgctl_albcs
    
[X1,Y1] = meshgrid(x,lat);
[X2,Y2] = meshgrid(kern_lon,kern_lat);
anom_clisccp_int=nan*ones(12,b,c,90,144);
avgctl_albcs_int=nan*ones(12,90,144);
for M=1:12
    avgctl_albcs_int(M,:,:) = interp2(X1,Y1,squeeze(cat_albcs(M,:,:)),X2,Y2);  
    for T=1:b
        for P=1:c
            anom_clisccp_int(M,T,P,:,:) = interp2(X1,Y1,squeeze(cat_clisccp(M,T,P,:,:)),X2,Y2);
        end
    end
end
clear cat_albcs cat_clisccp

%% Map each location's clear-sky surface albedo to the correct albedo bin
[a,b,c,d,e]=size(anom_clisccp_int); % (month,tau,CTP,lat,lon)
SWkernel_map=NaN*ones(12,b,c,d,e);
[X1,Y1] = meshgrid(albcs_midpt,1:b);
for M=1:a
    for LA=1:d
        alon=squeeze(avgctl_albcs_int(M,LA,:)); 
        if numel(alon(isnan(alon)))>0; continue; end
        [X2,Y2] = meshgrid(alon,1:b);
        for P=1:c
            SWkernel_map(M,:,P,LA,:)=interp2(X1,Y1,squeeze(SWkernel(M,:,P,LA,:)),X2,Y2);
        end
    end
end
LWkernel_map = repmat(LWkernel(:,:,:,:,1),[1,1,1,1,e]); % has no dependence on underlying albedo
clear LWkernel SWkernel

%% Calculate cloud feedback
% each feedback is size (MO,TAU,CTP,LAT,LON)=(12,7,7,90,144)
SW_cld_fdbk=anom_clisccp_int.*SWkernel_map/avgdtas;
LW_cld_fdbk=anom_clisccp_int.*LWkernel_map/avgdtas;

% ensure that SW_cld_fdbk is zero rather than NaN if the SW kernel is zero (e.g., in the polar night)
SW_cld_fdbk(SWkernel_map==0)=0;

%% Quick sanity check:
% print the global, annual mean LW and SW cloud feedbacks:
sumLW = squeeze(nanmean(nanmean(nansum(nansum(LW_cld_fdbk,2),3),1),5));
avgLW_cld_fbk = count_wt_mean(sumLW,kern_coslat,1);
display(['avg LW cloud feedback = ',num2str(avgLW_cld_fbk)])
sumSW = squeeze(nanmean(nanmean(nansum(nansum(SW_cld_fdbk,2),3),1),5));
avgSW_cld_fbk = count_wt_mean(sumSW,kern_coslat,1);
display(['avg SW cloud feedback = ',num2str(avgSW_cld_fbk)])

File: code/python/apply_cloud_kernels_v2.py
#!/usr/bin/env cdat
"""
% Cloud Property Histograms. Part I: Cloud Radiative Kernels. J. Climate, 25, 3715?3735. doi:10.1175/JCLI-D-11-00248.1.

% v2: This script is written to demonstrate how to compute the cloud feedback using for a 
% short (2-year) period of MPI-ESM-LR using the difference between amipFuture and amip runs.
% One should difference longer periods for more robust results -- these are just for demonstrative purposes

% Data that are used in this script:
% 1. model clisccp field
% 2. model rsuscs field
% 3. model rsdscs field
% 4. model tas field
% 5. cloud radiative kernels

% This script written by Mark Zelinka (zelinka1@llnl.gov) on 14 July 2017

"""

#IMPORT STUFF:
#=====================
import cdms2 as cdms
import cdutil
import MV2 as MV
import numpy as np
import pylab as pl

###########################################################################
# HELPFUL FUNCTIONS FOLLOW 
###########################################################################

###########################################################################
def add_cyclic(data):
    # Add Cyclic point around 360 degrees longitude:
    lons=data.getLongitude()[:]
    dx=np.gradient(lons)[-1]
    data2 = data(longitude=(0, dx+np.max(lons)), squeeze=True)    
    return data2

###########################################################################
def nanarray(vector):

    # this generates a masked array with the size given by vector
    # example: vector = (90,144,28)

    # similar to this=NaN*ones(x,y,z) in matlab

    this=MV.zeros(vector)
    this=MV.masked_where(this==0,this)

    return this

###########################################################################
def map_SWkern_to_lon(Ksw,albcsmap):

    from scipy.interpolate import interp1d
    ## Map each location's clear-sky surface albedo to the correct albedo bin
    # Ksw is size 12,7,7,lats,3
    # albcsmap is size A,lats,lons
    albcs=np.arange(0.0,1.5,0.5) 
    A=albcsmap.shape[0]
    TT=Ksw.shape[1]
    PP=Ksw.shape[2]
    lenlat=Ksw.shape[3]
    lenlon=albcsmap.shape[2]
    SWkernel_map=nanarray((A,TT,PP,lenlat,lenlon))
    for M in range(A):
        MM=M
        while MM>11:
            MM=MM-12
        for LA in range(lenlat):
            alon=albcsmap[M,LA,:] 
            # interp1d can't handle mask but it can deal with NaN (?)
            try:
                alon2=MV.where(alon.mask,np.nan,alon)   
            except:
                alon2=alon
            if np.ma.count(alon2)>1: # at least 1 unmasked value
                if len(pl.find(Ksw[MM,:,:,LA,:]>0))==0:
                    SWkernel_map[M,:,:,LA,:] = 0
                else:
                    f = interp1d(albcs,Ksw[MM,:,:,LA,:],axis=2)
                    ynew = f(alon2.data)
                    ynew=MV.masked_where(alon2.mask,ynew)
                    SWkernel_map[M,:,:,LA,:] = ynew
            else:
                continue

    return SWkernel_map

###########################################################################
# MAIN ROUTINE FOLLOWS
###########################################################################
direc='/work/zelinka1/git/cloud-radiative-kernels/data/'

# Load in the Zelinka et al 2012 kernels:
f=cdms.open(direc+'cloud_kernels2.nc')
LWkernel=f('LWkernel')
SWkernel=f('SWkernel')
f.close()

albcs=np.arange(0.0,1.5,0.5) # the clear-sky albedos over which the kernel is computed

# LW kernel does not depend on albcs, just repeat the final dimension over longitudes:
LWkernel_map=np.tile(np.tile(LWkernel[:,:,:,:,0],(1,1,1,1,1)),(144,1,1,1,1))(order=[1,2,3,4,0])

# Define the cloud kernel axis attributes
lats=LWkernel.getLatitude()[:]
lons=np.arange(1.25,360,2.5)
grid = cdms.createGenericGrid(lats,lons)

# Load in clisccp from two models
f=cdms.open(direc+'clisccp_cfMon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc','r')
clisccp1=f('clisccp')
f.close()
f=cdms.open(direc+'clisccp_cfMon_MPI-ESM-LR_amipFuture_r1i1p1_197901-198112.nc','r')
clisccp2=f('clisccp')
f.close()

# Make sure clisccp is in percent  
sumclisccp1=MV.sum(MV.sum(clisccp1,2),1)
sumclisccp2=MV.sum(MV.sum(clisccp2,2),1)   
if np.max(sumclisccp1) <= 1.:
    clisccp1 = clisccp1*100.        
if np.max(sumclisccp2) <= 1.:
    clisccp2 = clisccp2*100.

# Compute climatological annual cycle:
avgclisccp1=cdutil.ANNUALCYCLE.climatology(clisccp1) #(12, TAU, CTP, LAT, LON)
avgclisccp2=cdutil.ANNUALCYCLE.climatology(clisccp2) #(12, TAU, CTP, LAT, LON)
del(clisccp1,clisccp2)

# Compute clisccp anomalies
anomclisccp = avgclisccp2 - avgclisccp1

# Compute clear-sky surface albedo
f=cdms.open(direc+'rsuscs_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc','r')
rsuscs1 = f('rsuscs')
f.close()
f=cdms.open(direc+'rsdscs_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc','r')
rsdscs1 = f('rsdscs')
f.close()

albcs1=rsuscs1/rsdscs1
avgalbcs1=cdutil.ANNUALCYCLE.climatology(albcs1) #(12, 90, 144)
avgalbcs1=MV.where(avgalbcs1>1.,1,avgalbcs1) # where(condition, x, y) is x where condition is true, y otherwise
avgalbcs1=MV.where(avgalbcs1<0.,0,avgalbcs1)
del(rsuscs1,rsdscs1,albcs1)

# Load surface air temperature
f=cdms.open(direc+'tas_Amon_MPI-ESM-LR_amip_r1i1p1_197901-198112.nc','r')
tas1 = f('tas')
f.close()
f=cdms.open(direc+'tas_Amon_MPI-ESM-LR_amipFuture_r1i1p1_197901-198112.nc','r')
tas2 = f('tas')
f.close()

# Compute climatological annual cycle:
avgtas1=cdutil.ANNUALCYCLE.climatology(tas1) #(12, 90, 144)
avgtas2=cdutil.ANNUALCYCLE.climatology(tas2) #(12, 90, 144)
del(tas1,tas2)

# Compute global annual mean tas anomalies
anomtas = avgtas2 - avgtas1
avgdtas = cdutil.averager(MV.average(anomtas,axis=0), axis='xy', weights='weighted') # (scalar)

# Regrid everything to the kernel grid:
avgalbcs1 = add_cyclic(avgalbcs1)
avgclisccp1 = add_cyclic(avgclisccp1)
avgclisccp2 = add_cyclic(avgclisccp2)
avganomclisccp = add_cyclic(anomclisccp)
avgalbcs1_grd = avgalbcs1.regrid(grid,regridTool="esmf",regridMethod = "linear")
avgclisccp1_grd = avgclisccp1.regrid(grid,regridTool="esmf",regridMethod = "linear")
avgclisccp2_grd = avgclisccp2.regrid(grid,regridTool="esmf",regridMethod = "linear")
avganomclisccp_grd = avganomclisccp.regrid(grid,regridTool="esmf",regridMethod = "linear")

# Use control albcs to map SW kernel to appropriate longitudes
SWkernel_map = map_SWkern_to_lon(SWkernel,avgalbcs1_grd)

# Compute clisccp anomalies normalized by global mean delta tas
anomclisccp = avganomclisccp_grd/avgdtas

# Compute feedbacks: Multiply clisccp anomalies by kernels
SW0 = SWkernel_map*anomclisccp
LW_cld_fbk = LWkernel_map*anomclisccp
LW_cld_fbk.setAxisList(anomclisccp.getAxisList())

# Set the SW cloud feedbacks to zero in the polar night
# The sun is down if every bin of the SW kernel is zero:
sundown=MV.sum(MV.sum(SWkernel_map,axis=2),axis=1)  #12,90,144
repsundown=np.tile(np.tile(sundown,(1,1,1,1,1)),(7,7,1,1,1))(order=[2,1,0,3,4])
SW1 = MV.where(repsundown==0, 0, SW0) # where(condition, x, y) is x where condition is true, y otherwise
SW_cld_fbk = MV.where(repsundown.mask, 0, SW1) # where(condition, x, y) is x where condition is true, y otherwise
SW_cld_fbk.setAxisList(anomclisccp.getAxisList())

# SW_cld_fbk and LW_cld_fbk contain the contributions to the feedback from cloud anomalies in each bin of the histogram


# Quick sanity check:
# print the global, annual mean LW and SW cloud feedbacks:
sumLW = MV.average(MV.sum(MV.sum(LW_cld_fbk,axis=2),axis=1),axis=0)
avgLW_cld_fbk = cdutil.averager(sumLW, axis='xy', weights='weighted')
print 'avg LW cloud feedback = '+str(avgLW_cld_fbk)
sumSW = MV.average(MV.sum(MV.sum(SW_cld_fbk,axis=2),axis=1),axis=0)
avgSW_cld_fbk = cdutil.averager(sumSW, axis='xy', weights='weighted')
print 'avg SW cloud feedback = '+str(avgSW_cld_fbk)

# Some sample global mean figures
tau=[0.,0.3,1.3,3.6,9.4,23.,60.,380.]
ctp=[1000,800,680,560,440,310,180,50]

# amip cloud fraction histogram:
pl.subplots()
data = cdutil.averager(MV.average(avgclisccp1_grd,axis=0), axis='xy', weights='weighted').transpose()
pl.pcolor(data,shading='flat',cmap='Blues_r',vmin=0, vmax=10)
pl.xticks(np.arange(8), tau)
pl.yticks(np.arange(8), ctp)
pl.title('Global mean amip cloud fraction')
pl.xlabel('TAU')
pl.ylabel('CTP')
pl.colorbar()

# amipFuture cloud fraction histogram:
pl.subplots()
data = cdutil.averager(MV.average(avgclisccp2_grd,axis=0), axis='xy', weights='weighted').transpose()
pl.pcolor(data,shading='flat',cmap='Blues_r',vmin=0, vmax=10)
pl.xticks(np.arange(8), tau)
pl.yticks(np.arange(8), ctp)
pl.title('Global mean amipFuture cloud fraction')
pl.xlabel('TAU')
pl.ylabel('CTP')
pl.colorbar()

# difference of cloud fraction histograms:
pl.subplots()
data = cdutil.averager(MV.average(anomclisccp,axis=0), axis='xy', weights='weighted').transpose()
pl.pcolor(data,shading='flat',cmap='RdBu_r',vmin=-0.75, vmax=0.75)
pl.xticks(np.arange(8), tau)
pl.yticks(np.arange(8), ctp)
pl.title('Global mean normalized change in cloud fraction')
pl.colorbar()

# LW cloud feedback contributions:
pl.subplots()
data = cdutil.averager(MV.average(LW_cld_fbk,axis=0), axis='xy', weights='weighted').transpose()
pl.pcolor(data,shading='flat',cmap='RdBu_r',vmin=-0.75, vmax=0.75)
pl.xticks(np.arange(8), tau)
pl.yticks(np.arange(8), ctp)
pl.title('Global mean LW cloud feedback contributions')
pl.colorbar()

# SW cloud feedback contributions:
pl.subplots()
data = cdutil.averager(MV.average(SW_cld_fbk,axis=0), axis='xy', weights='weighted').transpose()
pl.pcolor(data,shading='flat',cmap='RdBu_r',vmin=-0.75, vmax=0.75)
pl.xticks(np.arange(8), tau)
pl.yticks(np.arange(8), ctp)
pl.title('Global mean SW cloud feedback contributions')
pl.colorbar()
pl.show()
